### ISSUE TITLE
Homepage SSL certificate expired?

### ISSUE THREAD

**Issue by carlocab** [Orc/discount#237]
I recently tried to upgrade Homebrew's version of discount here: https://github.com/Homebrew/homebrew-core/pull/66978

Homebrew CI runs a bunch of tests, and one of them apparently involves querying your homepage. It complained that the SSL certificate was expired.

Thought you'd like to know, just in case you hadn't noticed already.

**Comment by Orc** [comment]
That's bizarre, because my web browser hasn't said boo about it, despite the certificate being 47 days old?

What does the homebrew build scaffolding use to query the homepage?  wget, curl, or something else?

**Comment by carlocab** [comment]
It's a test using one of Homebrew's newer features, so I'm not very familiar with it. However, `curl` is the most likely candidate, as that's what Homebrew uses for pretty much everything else (and is readily available on macOS).

My browser hasn't complained about your website either. However:
```
❯ curl -I https://www.pell.portland.or.us/\~orc/Code/discount/
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```
Not really familiar with these protocols, so I can't say I understand what's going on. Also, for your references, this is the exact error message reported in CI:
```
==> brew livecheck discount --full-name --debug
==> FAILED
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/discount.rb

Formula:          discount
Livecheckable?:   Yes

URL:              https://www.pell.portland.or.us/~orc/Code/discount/
Strategy:         PageMatch
Regex:            /href=.*?discount[._-]v?(\d+(?:\.\d+)+[a-z]?)\.t/i
Error: discount: SSL_connect returned=1 errno=0 state=error: certificate verify failed (unable to get local issuer certificate)
```

**Comment by carlocab** [comment]
Ah, wait. Maybe I misunderstood. It's not a matter of an expired certificate after all...

**Comment by binki** [comment]
They’re not sending the entire certificate chain: https://www.ssllabs.com/ssltest/analyze.html?d=www.pell.portland.or.us  .

**Comment by Orc** [comment]
Ok, I checked further.   Apparently I can't send the entire certificate chain because the letsencrypt CA doesn't do delegated intermediate certificates (there's a certification process that -- at least according to commentary on the letsencrypt support group -- requires manual intervention, and that converts a basically free certification process into one that costs money.)

wget and curl, unlike web browsers, freak out if they have to fetch intermediate certificates.

So, unless some patron underwrites the cost of a certificate from a CA that can afford to delegate intermediate certificates, I can't afford ssl that makes curl happy.  I think your CI process either needs to fetch from the unencrypted http:// site or it needs to invoke curl with the equivalent of wget's **--no-check-certificate** option.

tl/dr;  Sorry, this is working as designed.  `¯\_(ツ)_/¯`

**Comment by binki** [comment]
> Ok, I checked further. Apparently I can't send the entire certificate chain because the letsencrypt CA doesn't do delegated intermediate certificates (there's a certification process that -- at least according to commentary on the letsencrypt support group -- requires manual intervention, and that converts a basically free certification process into one that costs money.)
> 
> wget and curl, unlike web browsers, freak out if they have to fetch intermediate certificates.

I think a fresh install of Firefox will freak out too unless if you had visited another website using the same intermediate certificate in the meantime. As you browse the web and visit sites which properly send the fully chain, it caches those certificates in its store. It seems like this is a common enough mistake in websites that many browsers ship a bunch of super-common intermediate certificates, such as the ones used by letsencrypt, to support visiting misconfigured websites. If they don’t do this, the user gets a bad experience and associates it with whatever browser brand they’re using at the moment ;-).

> So, unless some patron underwrites the cost of a certificate from a CA that can afford to delegate intermediate certificates, I can't afford ssl that makes curl happy. I think your CI process either needs to fetch from the unencrypted http:// site or it needs to invoke curl with the equivalent of wget's **--no-check-certificate** option.
> 
> tl/dr; Sorry, this is working as designed. `¯\_(ツ)_/¯`

I think you just need to point your webserver’s configuration at `fullchain.pem` instead of `cert.pem` if you actually have control over the webserver. Normal apache2/openssl/nginx just does the right thing if you do that IIRC ;-).

I just rechecked and it looks like you’re sending the full chain properly. `openssl s_client -showcerts -connect www.pell.portland.or.us:443` shows “Verification: OK” and multiple certificates now whereas it didn’t a few hours ago. So I think you fixed it without realizing it? Yay! ;-)

Thanks!

**Comment by Orc** [comment]
Maybe?  I updated lighttpd (apache eats processor like a black friday sale) to try to disable sslv3, and I guess it does something with intermediate certificates that the previous version I was using didn't?

**Comment by binki** [comment]
Interesting. The update probably fixed some configuration issue. Just glad it’s fixed (I might care about this particular issue a bit too much, sorry x.x)!

**Comment by carlocab** [comment]
So, I'm not really sure what happened (as you can tell; I don't really grok this stuff), but it's been fixed for me as well. And, even if it didn't, 

>tl/dr; Sorry, this is working as designed. `¯\_(ツ)_/¯`

I don't mind this answer at all either. Thanks for the time you spent looking into it!
